_drafts/Firmware-encryption-with-python.md

```diff
@@ -0,0 +1,643 @@
+```
+$ pip install pycryptodome
+```
+
+### Define encryption keys
+
+To define the encryption keys, we will generate a key to use in the script using the PyCryptodome library. This same key will also be used for decryption in the MCU. For our configuration, we will use AES-128, which requires a 16-byte key. Additionally, we will use a 2-byte Nonce value as the initialization vector.
+
+```python
+from Crypto import Random
+myKey = Random.get_random_bytes(16)
+myIV  = Random.get_random_bytes(2)
+print(myKey.hex())
+print(myIV.hex())
+```
+Here are the generated keys:
+
+```
+$ 583bf4901f623510a1408e8be7bd6f5a
+$ f824
+```
+
+### Encrypting with PyCryptodome
+
+Let’s use the generated keys to start encrypting data. The first step is to configure our counter block, which is a necessary parameter to create an object that allows us to encrypt using the `.encrypt()` method.
+
+#### Counter Block
+
+According to the PyCryptodome documentation, we can create a counter block by concatenating values before (prefix) and after (suffix) the counter variable:
+
+![](/img/fwup-encryption/Coun_block_1.png)
+
+Let’s use our key and Nonce value (AES_IVRx) to create the counter block using the `Counter.new()` method. The method requires several parameters:
+
+1. The desired counter length in bits.
+2. The initial value of the counter.
+3. The endianness of the counter data alignment.
+
+> counter length: 112 bit / 8 = 14 bytes
+> counter block length: 14 bytes + Nonce 2 bytes = 16 bytes = 128 bit
+
+Since we want a counter length of 112 bits and our Nonce is 2 bytes, the counter block length should be 16 bytes (128 bits) to match the AES-128 bit encryption. We want the counter variable to start at 0, so we’ll set initial_value=0. We’ll also specify little_endian=False for big-endian data alignment to align with the counter in the AES processor on the MCU:
+
+```python
+import binascii
+from Crypto.Cipher import AES
+from Crypto.Util import Counter
+myKey = "58:3b:f4:90:1f:62:35:10:a1:40:8e:8b:e7:bd:6f:5a"
+myIV  = "f8:24"
+AES_KEY = binascii.unhexlify(myKey.replace(":", ""))
+AES_NONCE = binascii.unhexlify(myIV.replace(":", ""))
+blockCounter = Counter.new(112, prefix=AES_NONCE, little_endian=False, initial_value=0)
+print( blockCounter )
+```
+We see the parameters of the counter block:
+
+```
+$ {'counter_len': 14, 'prefix': b'\xf8$', 'suffix': b'', 'initial_value': 0, 'little_endian': False}
+```
+The result should look something like this:
+
+![](/img/fwup-encryption/Coun_block_2.png)
+
+That’s a valid point. The size of the Nonce (PREFIX) affects the available space for the counter variable. If we have a larger Nonce value, it means less space will be available for the counter variable. For example, if we generate a 4-byte Nonce, the remaining 12 bytes will be allocated for the counter.
```

Comment:
What's a valid point? Not sure I follow here

Reply:
The intention is to mention that the larger the value used for Nonce will reduce the space for the counter variable.
Change the wording a bit.